File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/station/elevators/ElevatorStatusAdapter.java
package de.deutschebahn.bahnhoflive.ui.station.elevators;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import de.deutschebahn.bahnhoflive.backend.db.fasta2.model.FacilityStatus;
import de.deutschebahn.bahnhoflive.push.FacilityPushManager;
import de.deutschebahn.bahnhoflive.view.SingleSelectionManager;

abstract class ElevatorStatusAdapter extends RecyclerView.Adapter<FacilityStatusViewHolder> {

    private List<FacilityStatus> facilityStatuses;

    private final SingleSelectionManager selectionManager;

    private final FacilityPushManager facilityPushManager = FacilityPushManager.getInstance();

    ElevatorStatusAdapter() {
        selectionManager = new SingleSelectionManager(this);
        SingleSelectionManager.type = "d1_aufzuege";
    }

    @Override
    public FacilityStatusViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return onCreateViewHolder(parent, selectionManager, facilityPushManager);
    }

    @NonNull
    public abstract FacilityStatusViewHolder onCreateViewHolder(ViewGroup parent, SingleSelectionManager selectionManager, FacilityPushManager facilityPushManager);

    @Override
    public void onBindViewHolder(FacilityStatusViewHolder holder, int position) {
        holder.bind(facilityStatuses.get(position));
    }

    @Override
    public int getItemCount() {
        return facilityStatuses == null ? 0 : facilityStatuses.size();
    }

    public void setData(List<FacilityStatus> facilityStatuses) {
        this.facilityStatuses = new ArrayList<>(facilityStatuses);

        notifyDataSetChanged();
    }

    public void invalidateContent() {
        if (facilityStatuses != null) {
            notifyItemRangeChanged(0, facilityStatuses.size());
        }
    }

    public List<FacilityStatus> getData() {
        return facilityStatuses;
    }

    public FacilityStatus getSelectedItem() {
        return selectionManager.getSelectedItem(facilityStatuses);
    }
}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/ui/station/elevators/ElevatorStatusAdapter.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package de.deutschebahn.bahnhoflive.ui.station.elevators

import de.deutschebahn.bahnhoflive.push.FacilityPushManager.Companion.instance
import androidx.recyclerview.widget.RecyclerView
import de.deutschebahn.bahnhoflive.backend.db.fasta2.model.FacilityStatus
import de.deutschebahn.bahnhoflive.view.SingleSelectionManager
import de.deutschebahn.bahnhoflive.push.FacilityPushManager
import android.view.ViewGroup
import java.util.ArrayList

abstract class ElevatorStatusAdapter : RecyclerView.Adapter<FacilityStatusViewHolder>() {
    private var facilityStatuses: List<FacilityStatus>? = null
    private var selectionManager: SingleSelectionManager? = null
    private val facilityPushManager = instance

    init {
        selectionManager = SingleSelectionManager(this)
        SingleSelectionManager.type = "d1_aufzuege"
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FacilityStatusViewHolder {
        return onCreateViewHolder(parent, selectionManager, facilityPushManager)
    }

    abstract fun onCreateViewHolder(
        parent: ViewGroup,
        selectionManager: SingleSelectionManager?,
        facilityPushManager: FacilityPushManager
    ): FacilityStatusViewHolder

    override fun onBindViewHolder(holder: FacilityStatusViewHolder, position: Int) {
        holder.bind(facilityStatuses!![position])
    }

    override fun getItemCount(): Int {
        return if (facilityStatuses == null) 0 else facilityStatuses!!.size
    }

    open var data: List<FacilityStatus>?
        get() = facilityStatuses
        set(value) {
            this.facilityStatuses = value?.let { ArrayList(it) }
            notifyDataSetChanged()
        }

    val selectedItem: FacilityStatus?
        get() = selectionManager?.getSelectedItem(facilityStatuses)
}